### PR TITLE
Improve accessibility with more visible focus indication for non vue apps

### LINKED ISSUE #31239: [BITV/A11Y]: Header accessibility
* [x] Nextcloud logo should have more contrast on focus → we will change to thick outline
* [x] Right icons also don't have enough change between unfocused and focused
* [x] Increase contrast in app icons to at least 3:1

### PR DESCRIPTION
**Pull request by CarlSchwan**
- Add a focus effect on each header entry
image

- Add a focus effect on app navigation
image

- Add a focus effect on menus
image

- Add a focus effect on file app
image

- Add a focus effect on non vue input elements
image
 
- Show big focus outline when using focus-visible instead of focus
- Add polyfill for focus-visible since it's only very recently available on WebKit
- Add a few small screen readers improvements in the file app (checkbox)

I used https://www.sarasoueidan.com/blog/focus-indicators/ to make
sure the focus effect was WCAG 2 and WCAG 2.1 compliant.

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

### CODE REVIEW

File: core/css/apps.scss [hunk 1]
```diff
@@ -159,6 +159,13 @@ kbd {
 					background-color: var(--color-background-hover);
 				}
 			}
+			a:focus-visible {
+				border: 1px solid var(--color-primary);
```
Comment (jancborchardt) [inline core/css/apps.scss:163]: Small detail: Border thickness of 2px together with our standard `border-radius: var(--border-radius)` could look a bit better. :)

File: core/css/header.scss [hunk 2]
```diff
@@ -20,6 +20,9 @@
 	-webkit-user-select: none;
 	-moz-user-select: none;
 	-ms-user-select: none;
+	a:focus-visible, button:focus-visible, div[role="button"]:focus-visible {
+		outline: 1px solid var(--color-primary-text);
```
Comment (jancborchardt) [inline core/css/header.scss:24]: Or rather the 2px and border-radius is here? Or in both places? :)

File: core/templates/layout.user.php [hunk 3]
```diff
@@ -60,7 +60,7 @@
 					id="nextcloud">
 					<div class="logo logo-icon">
 						<h1 class="hidden-visually">
-							<?php p($theme->getName()); ?> <?php p(!empty($_['application'])?$_['application']: $l->t('Apps')); ?>
+							<?php p($l->t('$1\'s homepage', [$theme->getName()])); ?>
```
Comment (CarlSchwan) [inline core/templates/layout.user.php:63]:
```suggestion
							<?php p($l->t('%s\'s homepage', [$theme->getName()])); ?>
```

File: apps/files/css/files.scss [hunk 4]
```diff
@@ -1195,6 +1205,11 @@ table.dragshadow td.size {
 	#showgridview:focus + & {
 		opacity: 1;
 	}
+
+	&:focus-visible,
+	#showgridview:focus-visible + & {
+		outline: 2p2 solid var(--color-primary) !important;
```
Comment (vanpertsch) [inline apps/files/css/files.scss:1211]: little typo here

File: apps/files/templates/list.php [hunk 5]
```diff
@@ -44,24 +44,24 @@
...
+                        <a href="#" class="actions-selected">
                             <span class="icon icon-more"></span>
                             <span><?php p($l->t('Actions'))?></span>
                         </a>
 					</span>
 				</div>
 			</th>
 			<th id="headerSize" class="hidden column-size">
-				<a class="size sort columntitle" data-sort="size"><span><?php p($l->t('Size')); ?></span><span class="sort-indicator"></span></a>
+				<a class="size sort columntitle" href="#"data-sort="size"><span><?php p($l->t('Size')); ?></span><span class="sort-indicator"></span></a>
```
Comment (PVince81) [inline apps/files/templates/list.php:61]: missing space!!!

File: core/css/header.scss [hunk 6]
```diff
@@ -57,20 +62,18 @@
 
 /* LOGO and APP NAME -------------------------------------------------------- */
 #nextcloud {
-	padding: 7px 0;
+	padding: 5px 0;
 	padding-left: 86px; // logo width + 2* pa
 	position: relative;
-	height: 100%;
+	height: calc(100% - 4px);
```
Comment (skjnldsv) [inline core/css/header.scss:68]: Is this still needed? I guess it was because of the border you initially went for?

File: core/css/header.scss [hunk 7]
```diff
@@ -57,20 +62,18 @@
...
 	box-sizing: border-box;
 	opacity: 1;
 	align-items: center;
 	display: flex;
 	flex-wrap: wrap;
 	overflow: hidden;
+	margin: 2px;
```
Comment (skjnldsv) [inline core/css/header.scss:75]: same as above

File: core/css/header.scss [hunk 8]
```diff
@@ -583,7 +599,7 @@ nav[role='navigation'] {
 		border-width: 10px;
 		transform: translateX(-50%);
 		left: 50%;
-		bottom: 0;
+		bottom: -5px;
```
Comment (skjnldsv) [inline core/css/header.scss:602]: Seems odd, why is this needed?

**Review by skjnldsv — CHANGES_REQUESTED** [review]
See comments

**Comment by jancborchardt** [comment]
@CarlSchwan any update here? Would be good to get this in early in the 25 cycle then. :)

File: core/css/header.scss (hunk 6 above)
Comment (CarlSchwan) [reply core/css/header.scss:68]: yes this avoid the border touching the border of the websites

with

image

without

image

File: core/css/header.scss (hunk 8 above)
Comment (CarlSchwan) [reply core/css/header.scss:602]: This move the triangle back to the bottom of the header

With:

image

Without:

image

File: core/css/header.scss (hunk 7 above)
Comment (CarlSchwan) [reply core/css/header.scss:75]: this is to make sure the border is not completely bordering the window border, see screenshot above

File: apps/files/templates/list.php (hunk 5 above)
Comment (CarlSchwan) [reply apps/files/templates/list.php]: done :)

**Comment by PVince81** [comment]
some visual issues:

- [ ] bug: missing padding on left of the box, it's there for the other columns
image

- [ ] bug: border a bit cropped and also not fully centered:
image

apart from that it looks rather good!

I found some elements that are not focussable yet like the expanding arrow in the left sidebar, likely something for a separate PR as it's not directly related

**Comment by JuliaKirschenheuter** [comment]
Thank you for working on this issue and taking over accessibility topics! And sorry for late response!
I'm just trying to understand, to which issue (ticket) is this PR related to? There are much changes and i don't really know what should i actually check.

I think there are a lot of stuff which are related to all components (nc-vue). Sorry if i have missed something, but why this changes are not in nc-vue, but here? 

Have you fixed mostly visual stuff in this ticket? (or have you also adopted it to screen reader)? *not all buttons have (aria-label)
Could you explain, what this focus should cause (i haven't get it, sorry)?
image

"Shares" on the left sidebar have 2 stops.

Contacts have no focus
image

Some``` <li> ``` on a left sidebar are focusible with other design
image
same here in dropdown
image

"Toggle" have 2 stops
image


Focus of this dropdown menu will not go away with Esc (will not disappear)

image
image
image



And i think some of this visual issues would be better done in nc-vue? (@skjnldsv or i'm wrong in this point?) 

image
image
image
image

@CarlSchwan sorry if i checked "too much". It is great, that you work on it! I think it would be better to split this "elephant" improvements into smaller tickets - would be better to work on and to review! If you want - i could take over of a part of this tickets, you should't be alone =))

**Comment by JuliaKirschenheuter** [comment]
Thanks for your reply / fix!


> >Could you explain, what this focus should cause (i haven't get it, sorry)? image
> 
> It's mostly about keyboard navigation where having a focus effect visible is important. Screen reader navigation is a separate task unfortunately :(

Ok, sure. I meant here, that a user cannot interact with this focus area. I would suggest to stop on the next interaction element: "Shared".


> > "Shares" on the left sidebar have 2 stops.
> 
> Looks correct to me
> 
> image
> 
> image
> 

Ok, now for me too. I think a "shares" element should be somehow reworked in another PR, that it would have only 1 stop on interaction element (without whole "Shares", it makes not much sense).

> 
> > "Toggle" have 2 stops image
> 
> fixed

*still have 2 stops)

Lets move other stuff to separate tickets and do them one-by-one =)

**Review by JuliaKirschenheuter — APPROVED** [review]

**Comment by AndyScherzinger** [comment]
Resolves #31239

**Review by PVince81 — APPROVED** [review]
:+1: let's move forward

**Comment by CarlSchwan** [comment]
Need a rebase :) Going to do it today and then merge it

**Review by skjnldsv — APPROVED** [review]
